packages/loopring_v3/contracts/impl/relayer/FastWithdrawal.sol

```diff
@@ -0,0 +1,92 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+
+import "./RelayerData.sol";
+
+import "../../iface/IExchangeV3.sol";
+import "../../iface/ICookieJarContract.sol";
+
+import "../../lib/MathUint.sol";
+import "../../thirdparty/ECDSA.sol";
+
+/// @title An Implementation of IRelayer.
+/// @author Brecht Devos - <brecht@loopring.org>
+contract FastWithdrawal is RelayerData
+{
+    using ECDSA             for bytes32;
+    using MathUint          for uint;
+
+    function executeFastWithdrawal(
+        address cookieJarContractAddress,
+        address exchangeAddress,
+        address destinationAddress,
+        address from,
+        address to,
+        address token,
+        uint24  fAmount,
+        address feeToken,
+        uint16  fFee,
+        bool    onchainFeePayment,
+        uint32  salt,
+        bytes   calldata signature
+        )
+        external
+    {
+        // Check the signature
+        bytes32 hash = keccak256(
+            abi.encodePacked(
+                cookieJarContractAddress,
+                exchangeAddress,
+                destinationAddress,
+                from,
+                to,
+                token,
+                uint(fAmount).decodeFloat(24),
+                feeToken,
+                uint(fFee).decodeFloat(16),
+                onchainFeePayment,
+                salt
+            )
+        );
+        require(hash.toEthSignedMessageHash().recover(signature) == from, "UNAUTHORIZED");
```

Comment:
why not just check msg.sender == from?

Reply:
> For fast withdrawals/cookie jar use it is extremely important that only the owner of the cookie jar contract can actually execute the request. Or he has at least some guarantees from the operator. The owner of the cookie jar contract takes the risk that the user actually has the amount requested offchain AND that this amount will still be available when the conditional transfer is actually included in a block. So in practice the owner of the cookie jar and the operator will be the same entity or work closely together/come up with some kind of insurance so that this can be ensured. So a user can request a fast-withdrawal (with an ECDSA signed request in this PR) but the actual execution of the fast withdrawal (the onchain transfer) cannot be initiated by anyone other than the operator/cookie jar owner.

Essentially we don't know the offchain balance of a user onchain, so we can't allow a user to do a fast withdrawal of 100ETH if his offchain balance is only 1ETH.

Reply:
The user gives permission with a signature, the cookie jar owner gives permission by actually doing the fast withdrawal. In `CookieJarContract.transfer` this check is done:

```
// Only allow an authorized transaction origin to transfer tokens out
require(authorizedAddresses[tx.origin], "UNAUTHORIZED");
```

There are of course different ways to do this, but I think this would work okay.